### PR TITLE
Replace CombatRatingsAI.get_species_troops_grade with separate functions

### PR DESCRIPTION
**Pull request by Cjkjvfnby**
If you like it, I will make PR to partially merge this changes to master (so you can rebase without conflicts)

### CODE REVIEW

File: default/AI/freeorion_tools.py [hunk 1]
```diff
@@ -132,6 +132,25 @@ def chat_human(message):
...
+    Cache a function value by session.
+    Wrap only functions with hashable arguments.
+    """
+    _cache = {}
+    @wraps(function)
+    def wrapper(*args, **kwargs):
+        key = (function, args, tuple(kwargs.items()))
+        if key in _cache:
+            return _cache[key]
+        print 'cach miss'
```
Comment (Morlic-fo) [inline default/AI/freeorion_tools.py]: cache

**Comment by Morlic-fo** [comment]
Seems good to me.

**Comment by Cjkjvfnby** [comment]
Typo fixed.

OOps it is not really typo, it is missed debug print :) 

I move cache as separate PR to core: https://github.com/freeorion/freeorion/pull/585

**Comment by Cjkjvfnby** [comment]
I rebased my PR.

If you merge my branch in console before adding any commits, you will get fast-forward merge without merge commit.

**Pull request closed without merging**